frontend/src/metabase/public/containers/PublicQuestion/PublicQuestion.tsx

```diff
@@ -0,0 +1,245 @@
+  const [result, setResult] = useState<Dataset | null>(null);
+  const [initialized, setInitialized] = useState(false);
+  const [parameterValues, setParameterValues] = useState<ParameterValuesMap>(
+    {},
+  );
+
+  useMount(async () => {
+    if (uuid) {
+      setPublicQuestionEndpoints(uuid);
+    } else if (token) {
+      setEmbedQuestionEndpoints(token);
+    }
+
+    try {
+      let card;
+      if (token) {
+        card = await EmbedApi.card({ token });
+      } else if (uuid) {
+        card = await PublicApi.card({ uuid });
+      } else {
+        throw { status: 404 };
+      }
+
+      if (card.param_values) {
+        await dispatch(addParamValues(card.param_values));
+      }
+      if (card.param_fields) {
+        await dispatch(addFields(card.param_fields));
+      }
+
+      const parameters = getCardUiParameters(
+        card,
+        metadata,
+        {},
+        card.parameters || undefined,
+      );
+      const parameterValuesById = getParameterValuesByIdFromQueryParams(
+        parameters,
+        location.query,
+      );
+
+      setCard(card);
+      setParameterValues(parameterValuesById);
+    } catch (error) {
+      console.error("error", error);
+      dispatch(setErrorPage(error));
+    }
+  });
+
+  const setParameterValue = async (parameterId: ParameterId, value: any) => {
+    setParameterValues({
+      ...parameterValues,
+      [parameterId]: value,
+    });
+  };
+
+  const setParameterValueToDefault = (parameterId: ParameterId) => {
+    const parameters = getParameters();
+    const parameter = parameters.find(({ id }) => id === parameterId);
+    if (parameter) {
+      setParameterValue(parameterId, parameter.default);
+    }
+  };
+
+  const run = useCallback(async () => {
+    if (!card) {
+      return;
+    }
+
+    const parameters = card.parameters || getParametersFromCard(card);
+
+    try {
+      setResult(null);
+
+      let newResult;
+      if (token) {
+        // embeds apply parameter values server-side
+        newResult = await maybeUsePivotEndpoint(
+          EmbedApi.cardQuery,
+          card,
+        )({
+          token,
+          ...getParameterValuesBySlug(parameters, parameterValues),
+        });
+      } else if (uuid) {
+        // public links currently apply parameters client-side
+        const datasetQuery = applyParameters(card, parameters, parameterValues);
+        newResult = await maybeUsePivotEndpoint(
+          PublicApi.cardQuery,
+          card,
+        )({
+          uuid,
+          parameters: JSON.stringify(datasetQuery.parameters),
+        });
+      } else {
+        throw { status: 404 };
+      }
+
+      setResult(newResult);
+    } catch (error) {
+      console.error("error", error);
+      dispatch(setErrorPage(error));
+    }
+  }, [card, dispatch, parameterValues, token, uuid]);
+
+  useEffect(() => {
+    run();
+  }, [run]);
+
+  useEffect(() => {
+    if (result && !initialized) {
```

Comment:
technically there's no need to check for `!initialized` since you'll be setting the same value anyway. By doing this
```suggestion
    if (result) {
```

you reduce the number of times this effect runs. Also since you're setting the same value the component [won't be rerendered anyway](https://react.dev/reference/react/useState#setstate-caveats), so it's safe to call `setInitialized(true);` more than once.

> If the new value you provide is identical to the current state, as determined by an [Object.is](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) comparison, React will skip re-rendering the component and its children. This is an optimization. Although in some cases React may still need to call your component before skipping the children, it shouldn’t affect your code.

At the end of the day remove this condition cleans up the code just a bit. All the performance gain should be negligible

Reply:
oh yes. great catch, I had a passing thought of "wait a minute, we don't need this" but didn't want to go too crazy with touching code that I don't have a comprehensive view on. But this does make a ton of sense